clang/include/clang/Basic/TargetInfo.h

```diff
@@ -1764,6 +1764,13 @@ class TargetInfo : public TransferrableTargetInfo,
     return 0;
   }
 
+  /// \returns Target specific flat ptr address space; a flat ptr is a ptr that
+  /// can be casted to / from all other target address spaces. If the target
+  /// exposes no such address space / does not care, we return 0.
```

Comment:
This further spreads the notion that 0 is a lack of address space, which it is not. You shouldn't need a new thing for this; it should be equivalent to querying the target address space map for the LangAS Generic thing

Reply:
There's no such thing as LangAS::Generic, there's LangAS::Default. Default doesn't have to map to something that's actually useful, and it actually does not. There's no actual mechanism to get a flat pointer. And I have no idea what notion spreading you are talking about - this merely maintains the status quo, wherein everybody grabs an AS 0 pointer and hopes for the best. So I have no idea what you are talking about, if I'm honest.

Reply:
> here's no such thing as LangAS::Generic, there's LangAS::Default.

Right, so query the target address space map for that. 

The IR does not have the concept of a generic, flat, or no address space. 0 is the "default address space", which has its own specific properties. 

> this merely maintains the status quo, wherein everybody grabs an AS 0 pointer and hopes for the best.

For the most part this isn't true, at least in llvm proper. 

From the comment here, I'm not even sure what this is returning (either an IR address space or a clang address space)

Reply:
Some targets won't even have a flat address space. But that's fine, because in these specific uses the address space's underlying range/representation/meaning/whatever is completely irrelevant, all that matters is the GlobalObject (GlobalValue?) you get once you strip away an addrspacecasts. So I object to this on several grounds:

1. A flat address space may not exist
2. These uses don't care about the address space beyond being consistent across files
3. The existence of this API invites people to start using it for other things, and relying on properties of it that you've documented as being true which I don't think we want to be guaranteeing

If you really object to the arbitrary address space in use, I would rather see the special arrays be made overloaded/polymorphic instead so that the addrspacecasts can be avoided entirely .

Reply:
> Some targets won't even have a flat address space. But that's fine, because in these specific uses the address space's underlying range/representation/meaning/whatever is completely irrelevant, all that matters is the GlobalObject (GlobalValue?) you get once you strip away an addrspacecasts. So I object to this on several grounds:
> 
> 1. A flat address space may not exist
> 2. These uses don't care about the address space beyond being consistent across files
> 3. The existence of this API invites people to start using it for other things, and relying on properties of it that you've documented as being true which I don't think we want to be guaranteeing
> 
> If you really object to the arbitrary address space in use, I would rather see the special arrays be made overloaded/polymorphic instead so that the addrspacecasts can be avoided entirely .

@jrtc27 this is not purely bespoke for the special arrays (sadly). The core issue is the prevalent, unguarded/unassuming use of AS 0.  Since targets are not really bound to do anything special with 0 (there are no words anywhere saying 0 has any specific properties), problematic / invalid situations arise. This leads to a perpetual game of whack-a-mole as e.g. one brings up a more general purpose language, which is not quite ideal.

Strictly regarding the special purpose arrays, I object to the arbitrary address space because it actually breaks targets as things stand today, for example: SPIR-V uses 0 for private, if we just use 0 we end up trying to generate invalid SPIR-V that casts from global (CrossWorkgroup) to private (Function) when populating them. Yes we can then go and add special handling just for the special arrays etc. in the SPIR-V BE/the downstream Translator, but that seems less than ideal to me.

I'd like to address (1) and (3) (I did worry about those, but we have an interface that returns a vtbl ptr address space, and it was not clear why 0, which on some targets points to a limited visibility/capacity/private/unambiguously non-global private AS, would be valid for a vtable) - instead of unconditionally returning an integer, we could return an `optional`, with the default being to return `std::nullopt`. This'd constrain to uses to folks who know what they want, and why they want it, as opposed to trivial drop-in replacement.

Reply:
> > here's no such thing as LangAS::Generic, there's LangAS::Default.
> 
> Right, so query the target address space map for that.
> 
Ok, I queried LangAS::Default and I got [this](https://github.com/llvm/llvm-project/blob/08de0b3cf54e4998799673f835e9a7d5ead8efab/clang/lib/Basic/Targets/AMDGPU.cpp#L65) because of [this](https://github.com/llvm/llvm-project/blob/08de0b3cf54e4998799673f835e9a7d5ead8efab/clang/lib/Basic/Targets/AMDGPU.cpp#L256). What's next?

> The IR does not have the concept of a generic, flat, or no address space. 0 is the "default address space", which has its own specific properties.
> 

What are these specific properties and where are they documented? Also, 0 doesn't quite look like the default address space, at least not de jure, see above; it's not even the default address space for all uses of AMDGPU.

> > this merely maintains the status quo, wherein everybody grabs an AS 0 pointer and hopes for the best.
> 
> For the most part this isn't true, at least in llvm proper.
>

Sure, but this is in Clang, which pretty liberally uses things like `getUnqualPtrTy` and hopes it will work. Elsewhere you are essentially arguing for exactly the same thing, namely the FE should just use 0 unconditionally. I am saying that is wrong in general, and the next best thing is to query the target if it happens to have any opinions regarding what the AS one should use everywhere is. Just like we're already asking if it has any opinions about something far more niche, namely vtable ptrs), and then revert to the status quo which is use 0. In Clang. Because, otherwise, your request for the FE to "just do the right thing" ends up unsolvable.